internal/graph/track_check_resolver.go

```diff
@@ -0,0 +1,208 @@
+	ticker   *time.Ticker
+	logger   logger.Logger
+	limiter  *rate.Limiter
+	ctx      context.Context
+	nodes    sync.Map
+}
+
+var _ CheckResolver = (*TrackerCheckResolver)(nil)
+
+func WithTrackerLogger(logger logger.Logger) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.logger = logger
+	}
+}
+
+func WithTrackerContext(ctx context.Context) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.ctx = ctx
+	}
+}
+
+// Expired check the current tuple entry expiration.
+func (r *resolutionTree) expired() bool {
+	return time.Since(r.tm) > trackerInterval
+}
+
+// NewTrackCheckResolver creates an instance tracker Resolver.
+func NewTrackCheckResolver(opts ...TrackerCheckResolverOpt) *TrackerCheckResolver {
+	t := &TrackerCheckResolver{
+		limiter: rate.NewLimiter(rate.Limit(trackerLogLines/trackerLogInterval), trackerLogBurst),
+		ticker:  time.NewTicker(trackerInterval),
+	}
+
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	t.delegate = t
+	t.launchFlush()
+	return t
+}
+
+// LogExecutionPaths reports the model and tuple path.
+func (t *TrackerCheckResolver) logExecutionPaths(flush bool) {
+	t.nodes.Range(func(k, v any) bool {
+		modelid := k.(string)
+		paths, _ := v.(*sync.Map)
+		paths.Range(func(k, v any) bool {
+			tree, _ := v.(*resolutionTree)
+			path := k.(string)
+			if tree.expired() || flush {
+				if t.limiter.Allow() {
+					t.logger.Info("execution path hits",
+						zap.String("model", modelid),
+						zap.String("path", path),
+						zap.Uint64("hits", tree.hits.Load()))
+					paths.Delete(path)
+				}
+			}
+			return true
+		})
+		return true
+	})
+}
+
+// LaunchFlush starts the execution path logging and removal of entries.
+func (t *TrackerCheckResolver) launchFlush() {
+	go func() {
+		for {
+			select {
+			case <-t.ctx.Done():
+				t.ticker.Stop()
+				t.logExecutionPaths(true)
+				return
+			case <-t.ticker.C:
+				t.logExecutionPaths(false)
+				t.ticker.Reset(trackerInterval)
+			}
+		}
+	}()
+}
+
+// SetDelete assigns the next deletegate in the chain.
+func (t *TrackerCheckResolver) SetDelegate(delegate CheckResolver) {
+	t.delegate = delegate
+}
+
+// GetDelegate return the assigned delegate.
+func (t *TrackerCheckResolver) GetDelegate() CheckResolver {
+	return t.delegate
+}
+
+// Close implements CheckResolver.
+func (*TrackerCheckResolver) Close() {}
+
+// UserType returns the associated tuple user type.
+func (t *TrackerCheckResolver) userType(userKey string) string {
+	return string(tuple.GetUserTypeFromUser(userKey))
+}
+
+// GetTK returns formatted tuple suitable insertion into list.
+func (t *TrackerCheckResolver) getTK(tk *openfgav1.TupleKey) string {
+	return fmt.Sprintf("%s#%s@%s", tk.GetObject(), tk.GetRelation(), t.userType(tk.GetUser()))
+}
+
+// LoadModel populate model id for individual tuple paths.
+func (t *TrackerCheckResolver) loadModel(r *ResolveCheckRequest) (value any, ok bool) {
+	model := r.GetAuthorizationModelID()
+	value, ok = t.nodes.Load(model)
```

Comment:
The key to the map cannot be the model ID alone. There is no guarantee that a model ID is unique across stores. Please also add store ID as part of the key

Reply:
👍